templates/fine-tune-llm_v2/cookbooks/optimize_cost/README.ipynb

```diff
@@ -0,0 +1,302 @@
+    "\\* subject to availability.\n",
+    "\n",
+    "\n",
+    "For instance if I want to use the example config `./training_configs/lora/llama-3-70b.yaml` and run a 70B model on A100-80G I have to change the following parts:\n",
+    "\n",
+    "```\n",
+    "num_devices: 8\n",
+    "trainer_resources: {}\n",
+    "worker_resources:\n",
+    "  accelerator_type:A100-80G: 0.001\n",
+    "```\n",
+    "\n",
+    "All of our default configs are setup for A10G machines due to their more accesibility.\n",
+    "\n",
+    "#### Difference between `worker_resources` and `trainer_resources`\n",
+    "\n",
+    "In LoRA training you can often ignore `trainer_resources` and just provide the GPU type required for doing training under `worker_resources` similar to the example above.\n",
+    "\n",
+    "However, for full-parameter training on hetergoneous clusters (e.g. training 70B on A10G GPUs) it is a bit more convoluted. This is generally not recommended, but when you do not have access to X100 machines this is the only way you can still train super large models (fp16 and full-parameter).\n",
+    "\n",
+    "Ray Train allows users to specify a different set of resources for rank-0 vs. other ranks. Rank-0 is responsible for checkpoining and normally needs more CPU RAM than the other workers at the time of checkpointing, because of the implementation details around weight aggregation. In a heterogenous multi-node setting where you have both small nodes and large nodes with the same GPUs this can cause a problem, because it becomes important where rank-0 is stored and the cluster is not symetrically used all the time. \n",
+    "\n",
+    "A prime example is running fine-tuning on `A10G`s. On AWS `A10s` are available in `g5.4xlarge` with 1 GPU and small RAM capacity all the way to `g5.48xlarge` with 8 GPUs and large RAM. During checkpoining a large model like 70B, the CPU RAM on g5.4xlarge is not sufficient and hence we have to define `memory` requirement for `trained_resources` to ensure that the large instance gets picked for rank-0. For example let's look at `./training_configs/full_param/llama-3-70b.yaml`:\n",
+    "\n",
+    "```\n",
+    "num_devices: 32 \n",
+    "# Head node would have at least 200 GB memory\n",
+    "trainer_resources:\n",
+    "  memory: 161061273600 # 130 GB memory\n",
+    "worker_resources:\n",
+    "  memory: 53687091200 # 50 GB memory\n",
+    "  accelerator_type:A10G: 0.001\n",
+    "```\n",
+    "\n",
+    "This configs is asking for 32xA10G GPUs but does not specify the architecture (e.g. `g5.4xlarge` vs. `g5.48xlarge`). However, it specifies that there should be at least 130G of memory available for rank-0, which forces the Anyscale's instance manager to pick g5.48xlarge for rank-0 and any other A10s for other workers hence not hitting RAM OOM issues during checkpointing. This is much better than being forced to use `4xg5.48xlarge` nodes. They may not be available on-demand and might be more expensive too."
```

Comment:
Is it worth adding a small note here on how Ray/Anyscale will chose worker nodes when multiple GPUs are requested? Just something like  "The Ray autoscaler will prioritize instances with the highest number of GPUs. If there is insufficient capacity, it will proceed to select instances in decreasing order of their GPU count."

Feel like it's better to mention in case a person is surprised that they got 2x nodes with 2-A10s instances instead of 1-4xA10s.